### PR TITLE
Connection tracking support

### PR DESCRIPTION
**Pull request by usha1830**
Connection tracking support for PNA needs the following extern functions to be supported.
- `add_entry `extern with additional timeout parameter
- `restart_expire_timer `
- `set_entry_expire_timer`

DPDK provides the following corresponding instructions for connection tracking

- ` learn `instruction with timeout parameter (The second parameter of this instruction "action_params" is optional in DPDK)
- `rearm` instruction 
- `rearm` instruction with timeout parameter

The timeout parameter is the index in timeout structure defined in the learner table.

All of the above instruction expect all the arguments in metadata and hence if any argument is constant, it should be moved to metadata.

These externs have constraints on where these externs can be called from. Added a new pass to check all these constraints.
The existing convertToDpdk pass is enhanced to process these new externs and generate equivalent assembly spec.

DPDK reference patch describing the rearm and learn instructions:
https://patches.dpdk.org/project/dpdk/patch/20220421155951.31811-3-cristian.dumitrescu@intel.com/

### CODE REVIEW

File: backends/dpdk/dpdkArch.cpp [hunk 1]
```diff
@@ -2245,12 +2251,83 @@ void CollectAddOnMissTable::postorder(const IR::MethodCallStatement *mcs) {
...
+            if (add_on_miss != nullptr) {
+                if (add_on_miss->value->is<IR::ExpressionValue>()) {
+                    auto expr = add_on_miss->value->to<IR::ExpressionValue>()->expression;
+                    if (!expr->is<IR::BoolLiteral>()) {
+                        ::error(ErrorType::ERR_UNEXPECTED,
+                                "%1%: expected boolean for 'add_on_miss' property", add_on_miss);
+                        return;
+                    } else {
+                        auto use_add_on_miss = expr->to<IR::BoolLiteral>()->value;
+                        if (use_add_on_miss){
```
Comment (github-sajan) [inline backends/dpdk/dpdkArch.cpp]: cpplint error for no space ?

File: backends/dpdk/dpdkContext.cpp [hunk 2]
```diff
@@ -236,16 +270,17 @@ DpdkContextGenerator::addMatchAttributes(const IR::P4Table*table, const cstring
...
                 if (param->type->is<IR::Type_Bits>()) {
-                    addImmediateField(immFldArray, param->name.originalName,
-                                      index/8, param->type->width_bits());
-                    index += param->type->width_bits();
-                    position++;
-                } else {
+                    param_width = param->type->width_bits();
+                } else if (!param->type->is<IR::Type_Boolean>()) {
                     BUG("Unsupported parameter type %1%", param->type);
                 }
```
Comment (github-sajan) [inline backends/dpdk/dpdkContext.cpp:279]: Earlier error was thrown for param type other than Type_Bits. Now should error be limited to only Type_Boolean ?

File: backends/dpdk/dpdkHelpers.cpp [hunk 3]
```diff
@@ -768,26 +768,77 @@ bool ConvertStatementToDpdk::preorder(const IR::MethodCallStatement *s) {
...
             auto action = a->expr->arguments->at(0)->expression;
             auto action_name = action->to<IR::StringLiteral>()->value;
             auto param = a->expr->arguments->at(1)->expression;
+            if (argSize == 3)
+                timeout_id = a->expr->arguments->at(2)->expression;
+            if (timeout_id->is<IR::Constant>() ||  argSize == 2) {
+                // FIXME This code is valid if specification defines timeout as optional param.
+                BUG_CHECK(metadataStruct, "Metadata structure missing unexpectedly!");
+                IR::ID tmo(refmap->newName("timeout_id"));
+                if (argSize == 3)
```
Comment (github-sajan) [inline backends/dpdk/dpdkHelpers.cpp]: Can line 788 and line 794 if block be written under one if condition?

File: backends/dpdk/dpdkContext.cpp (hunk 2 above)
Comment (usha1830) [reply backends/dpdk/dpdkContext.cpp:279]: No error for Type_Bits and Type_Boolean. Other types are unsupported

File: backends/dpdk/dpdkArch.cpp (hunk 1 above)
Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp]: Updated.

File: backends/dpdk/dpdkHelpers.cpp (hunk 3 above)
Comment (usha1830) [reply backends/dpdk/dpdkHelpers.cpp]: Updated

File: backends/dpdk/dpdkContext.cpp (hunk 2 above)
Comment (github-sajan) [reply backends/dpdk/dpdkContext.cpp:279]: Okay. got it. The minimum param width is 8 in dpdk for any param type. Looks fine

**Review by github-sajan — APPROVED** [review]
Looks fine to me

File: backends/dpdk/spec.cpp [hunk 4]
```diff
@@ -449,13 +460,16 @@ std::ostream& IR::DpdkLearner::toSpec(std::ostream& out) const {
...
     }
-    if (auto size = properties->getProperty("psa_idle_timeout")) {
-        out << "\ttimeout " << DPDK::toStr(size->value) << "" << std::endl;
-    } else {
-        out << "\ttimeout " << DEFAULT_LEARNER_TABLE_TIMEOUT << std::endl;
-    }
 
-    out << "}" << std::endl;
+    // The timeout values are currently hardcoded here
+    // FIXME: This should come from P4 program but currently we don't have a way
```
Comment (usha1830) [inline backends/dpdk/spec.cpp]: This is decided to be a some default hardcoded value. The values and number of values in the timeout structure are yet to be confirmed. Will remove this FIXME once Cristian confirms the default vaues

Comment (usha1830) [reply backends/dpdk/spec.cpp]: This is updated after confirmation from Cristian.

File: backends/dpdk/dpdkHelpers.h [hunk 5]
```diff
@@ -176,6 +176,7 @@ class ConvertStatementToDpdk : public Inspector {
     void process_logical_operation(const IR::Expression*, const IR::Operation_Binary*);
     void process_relation_operation(const IR::Expression*, const IR::Operation_Relation*);
     cstring append_parser_name(const IR::P4Parser* p, cstring);
+//    bool checkTblCtxtValid(cstring, const IR::P4Action *);
```
Comment (kamleshbhalui) [inline backends/dpdk/dpdkHelpers.h:179]: can we get rid of this?

**Review by kamleshbhalui — APPROVED** [review]

**Comment by usha1830** [comment]
Merged with open source . Closing this PR

**Pull request closed without merging**